Index: package/hostapd/files/wpa_supplicant.sh
===================================================================
--- package/hostapd/files/wpa_supplicant.sh	(revision 26177)
+++ package/hostapd/files/wpa_supplicant.sh	(working copy)
@@ -49,21 +49,43 @@
 			esac
 		;;
 		*psk*)
-			key_mgmt='WPA-PSK'
 			config_get_bool usepassphrase "$vif" passphrase 1
 			if [ "$usepassphrase" = "1" ]; then
 				passphrase="psk=\"${key}\""
 			else
 				passphrase="psk=${key}"
 			fi
-			case "$enc" in
-				*psk2*)
-					proto='proto=RSN'
-					config_get ieee80211w "$vif" ieee80211w
-				;;
-				*psk*)
+
+			config_get vifmode "$vif" mode
+			case "$vifmode" in
+				adhoc)
+					key_mgmt='WPA-NONE'
+					pairwise='pairwise=NONE'
 					proto='proto=WPA'
+					ap_scan='ap_scan=2'
+					mode='mode=1'
+					case "$enc" in
+						*psk2*|*ccmp*|*aes*)
+							group='group=CCMP'
+						;;
+						*psk*|*tkip*)
+							group='group=TKIP'
+						;;
+					esac
 				;;
+				*)
+					key_mgmt='WPA-PSK'
+					case "$enc" in
+						*psk2*)
+							proto='proto=RSN'
+							config_get ieee80211w "$vif" ieee80211w
+						;;
+						*psk*)
+							proto='proto=WPA'
+						;;
+					esac
+				;;
+
 			esac
 		;;
 		*wpa*|*8021x*)
@@ -109,9 +131,11 @@
 	rm -rf /var/run/wpa_supplicant-$ifname
 	cat > /var/run/wpa_supplicant-$ifname.conf <<EOF
 ctrl_interface=/var/run/wpa_supplicant-$ifname
+$ap_scan
 network={
 	scan_ssid=1
 	ssid="$ssid"
+	$mode
 	$bssid
 	key_mgmt=$key_mgmt
 	$proto
@@ -133,6 +157,11 @@
 	$wep_tx_keyidx
 }
 EOF
-	[ -z "$proto" -a "$key_mgmt" != "NONE" ] || \
-		wpa_supplicant ${bridge:+ -b $bridge} -B -P "/var/run/wifi-${ifname}.pid" -D ${driver:-wext} -i "$ifname" -c /var/run/wpa_supplicant-$ifname.conf $options
+	[ -z "$proto" -a "$key_mgmt" != "NONE" ] || {
+		if [ "${driver:-wext}" == "madwifi" ]; then
+			wpa_supplicant ${bridge:+ -b $bridge} -B -P "/var/run/wifi-${ifname}.pid" -i "$ifname" -c /var/run/wpa_supplicant-$ifname.conf $options
+		else
+			wpa_supplicant ${bridge:+ -b $bridge} -B -P "/var/run/wifi-${ifname}.pid" -D ${driver:-wext} -i "$ifname" -c /var/run/wpa_supplicant-$ifname.conf $options
+		fi
+	}
 }
Index: package/madwifi/files/lib/wifi/madwifi.sh
===================================================================
--- package/madwifi/files/lib/wifi/madwifi.sh	(revision 26177)
+++ package/madwifi/files/lib/wifi/madwifi.sh	(working copy)
@@ -360,8 +360,8 @@
 					}
 				fi
 			;;
-			wds|sta)
-				if eval "type wpa_supplicant_setup_vif" 2>/dev/null >/dev/null; then
+			wds|sta|adhoc)
+				if [ -n "$start_hostapd" ] && eval "type wpa_supplicant_setup_vif" 2>/dev/null >/dev/null; then
 					wpa_supplicant_setup_vif "$vif" madwifi || {
 						echo "enable_atheros($device): Failed to set up wpa_supplicant for interface $ifname" >&2
 						ifconfig "$ifname" down
